### PR TITLE
support letsencrypt

### PR DESCRIPTION
**Pull request by ulrikstrid**
Created to allow this to happen: https://github.com/mmaker/ocaml-letsencrypt/pull/31

### CODE REVIEW

File: jose/Header.ml [hunk 1]
```diff
@@ -2,17 +2,26 @@ open Utils
...
   alg : Jwa.alg;
-  jku : string option;
   jwk : Jwk.public Jwk.t option;
   kid : string option;
   x5t : string option;
   x5t256 : string option;
   typ : string option;
   cty : string option;
   enc : Jwa.enc option;
+  extra : (string * Yojson.Safe.t) list option;
```
Comment (anmonteiro) [inline jose/Header.ml:12]: why does this have to be an option type? Wouldn't an empty list do the trick?

File: jose/Header.ml [hunk 2]
```diff
@@ -21,28 +30,41 @@ let make_header ?typ ?alg ?enc (jwk : Jwk.priv Jwk.t) =
...
-    jwk = None;
-    kid = Jwk.get_kid jwk;
+    jwk = (if jwk_header then Some (Jwk.pub_of_priv jwk) else None);
+    kid;
     x5t = None;
     x5t256 = None;
     typ;
     cty = None;
     enc;
+    extra = (match extra with [] -> None | extra -> Some extra);
```
Comment (anmonteiro) [inline jose/Header.ml:51]: see my comment above. this can just be a list instead of an option type.

File: jose/Jwk.ml [hunk 3]
```diff
@@ -37,6 +37,24 @@ module Util = struct
            let k = Mirage_crypto_ec.P256.Dsa.pub_of_cstruct point in
            k |> U_Result.get_exn)
 
+  let get_ES384_x_y key =
+    let point = Mirage_crypto_ec.P384.Dsa.pub_to_cstruct key in
+    let x_cs, y_cs = Cstruct.(split (shift point 1) 48) in
```
Comment (anmonteiro) [inline jose/Jwk.ml:42]: probably needs a comment why it's 48: 384 / 8 = 48

File: jose/Jwk.ml [hunk 4]
```diff
@@ -272,6 +348,23 @@ let to_priv_pem (jwk : priv t) =
       Ok (X509.Private_key.encode_pem (`RSA rsa.key) |> Cstruct.to_string)
   | _ -> Error `Not_rsa
 
+let of_priv_x509 ?use x509 : (priv t, [> `Not_rsa ]) result =
+  match x509 with
+  | `RSA priv_key -> Ok (make_priv_rsa ?use priv_key)
+  | `P256 priv_key -> Ok (make_priv_es256 ?use priv_key)
+  | `P384 priv_key -> Ok (make_priv_es384 ?use priv_key)
+  | `P521 priv_key -> Ok (make_priv_es512 ?use priv_key)
+  | _ -> Error (`Msg "key type not supported")
```
Comment (anmonteiro) [inline jose/Jwk.ml:357]: isn't this `` `Unsupported_kty``?

File: jose/Jwk.ml [hunk 5]
```diff
@@ -272,6 +348,23 @@ let to_priv_pem (jwk : priv t) =
...
+
+let of_pub_x509 ?use (x509 : X509.Public_key.t) :
+    (public t, [> `Not_rsa ]) result =
+  match x509 with
+  | `RSA public_key -> Ok (make_pub_rsa ?use public_key)
+  | `P256 public_key -> Ok (make_pub_es256 ?use public_key)
+  | `P384 public_key -> Ok (make_pub_es384 ?use public_key)
+  | `P521 public_key -> Ok (make_pub_es512 ?use public_key)
+  | _ -> Error (`Msg "key type not supported")
```
Comment (anmonteiro) [inline jose/Jwk.ml:366]: isn't this `` `Unsupported_kty ``?

File: jose/Jws.ml [hunk 6]
```diff
@@ -19,10 +21,72 @@ let of_string token =
...
+            raw_header = protected;
+            payload;
+            signature;
+          }
+    | Error e, _ -> Error e
+    | _, None -> Error `Not_supported
+  with _ -> Error `Not_json
+
+let of_string token =
+  match of_json_string token with
```
Comment (anmonteiro) [inline jose/Jws.ml:47]: isn't there another heuristic we can use to check if this is compact or general than trying to parse the whole thing?

Comment (ulrikstrid) [reply jose/Jws.ml:47]: I guess we can check if it starts with `ey` or `{`

**Pull request merged**